In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
y2015 = pd.read_csv('LoanStats3d.csv',skipinitialspace=True, header=1)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
y2015 = y2015[:-2]
y2015.dropna(axis=1, thresh=20000,inplace=True)

In [4]:
y2015.tail()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
421090    10000.0      10000.0          10000.0   36 months   11.99%   
421091    24000.0      24000.0          24000.0   36 months   11.99%   
421092    13000.0      13000.0          13000.0   60 months   15.99%   
421093    12000.0      12000.0          12000.0   60 months   19.99%   
421094    20000.0      20000.0          20000.0   36 months   11.99%   

        installment grade sub_grade                        emp_title  \
421090       332.10     B        B5                 Office Assistant   
421091       797.03     B        B5          Director of Maintenance   
421092       316.07     D        D2         Radiologist Technologist   
421093       317.86     E        E3                          Painter   
421094       664.20     B        B5  Manager Hotel Operations Oasis    

       emp_length         ...          percent_bc_gt_75  pub_rec_bankruptcies  \
421090    8 years         ...                     100.0                   0.0   
421091  10+ years         ...                     100.0                   0.0   
421092    5 years         ...                      50.0                   1.0   
421093     1 year         ...                      66.7                   0.0   
421094  10+ years         ...                      50.0                   0.0   

       tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
421090       0.0         32950.0           25274.0         9200.0   
421091       0.0        152650.0            8621.0         9000.0   
421092       0.0         51239.0           34178.0        10600.0   
421093       0.0         96919.0           58418.0         9700.0   
421094       1.0         43740.0           33307.0        41700.0   

       total_il_high_credit_limit hardship_flag disbursement_method  \
421090                    15850.0             N                Cash   
421091                        0.0             N                Cash   
421092                    33239.0             N                Cash   
421093                    69919.0             N                Cash   
421094                        0.0             N                Cash   

       debt_settlement_flag  
421090                    N  
421091                    N  
421092                    N  
421093                    N  
421094                    N  

[5 rows x 106 columns]

In [5]:
y2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421095 entries, 0 to 421094
Columns: 106 entries, loan_amnt to debt_settlement_flag
dtypes: float64(81), object(25)
memory usage: 340.5+ MB


In [6]:
categorical = y2015.select_dtypes(include=['object'])
for i in categorical:
    column = categorical[i]
    print(i)
    print(column.nunique())

term
2
int_rate
111
grade
7
sub_grade
35
emp_title
120812
emp_length
11
home_ownership
4
verification_status
3
issue_d
12
loan_status
7
pymnt_plan
2
purpose
14
title
27
zip_code
914
addr_state
49
earliest_cr_line
668
revol_util
1211
initial_list_status
2
last_pymnt_d
50
next_pymnt_d
6
last_credit_pull_d
51
application_type
2
hardship_flag
2
disbursement_method
1
debt_settlement_flag
2


In [7]:
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

y2015.drop(['emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state'], 1, inplace=True)

In [8]:
pd.get_dummies(y2015)

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
0         35000.0      35000.0          35000.0     14.85       829.90   
1         10400.0      10400.0          10400.0     22.45       289.91   
2         20000.0      20000.0          20000.0     10.78       432.66   
3          3600.0       3600.0           3600.0     13.99       123.03   
4         24700.0      24700.0          24700.0     11.99       820.28   
5         11950.0      11950.0          11950.0     13.44       405.18   
6         16000.0      16000.0          16000.0     12.88       363.07   
7         20000.0      20000.0          20000.0      9.17       637.58   
8          1400.0       1400.0           1400.0     12.88        47.10   
9         18000.0      18000.0          18000.0     19.48       471.70   
10        22400.0      22400.0          22400.0     12.88       508.30   
11        10000.0      10000.0          10000.0      6.49       306.45   
12         8000.0       8000.0           8000.0     11.48       263.74   
13        20000.0      20000.0          20000.0      8.49       631.26   
14        28000.0      28000.0          28000.0      6.49       858.05   
15        23000.0      23000.0          23000.0      8.49       471.77   
16         8650.0       8650.0           8650.0     19.89       320.99   
17        29900.0      29900.0          29900.0     12.88       678.49   
18         9600.0       9600.0           9600.0      7.49       298.58   
19        16000.0      16000.0          16000.0     14.85       379.39   
20        18000.0      18000.0          18000.0     11.99       400.31   
21        25000.0      25000.0          25000.0      7.49       777.55   
22        16000.0      16000.0          16000.0     12.88       363.07   
23         4200.0       4200.0           4200.0     13.99       143.53   
24        12000.0      12000.0          12000.0      9.17       382.55   
25         4000.0       4000.0           4000.0      9.17       127.52   
26        14000.0      14000.0          14000.0      9.80       450.43   
27        10500.0      10500.0          10500.0     15.77       367.96   
28        10000.0      10000.0          10000.0     11.48       329.67   
29        15000.0      15000.0          15000.0      9.80       482.61   
...           ...          ...              ...       ...          ...   
421065     4200.0       4200.0           4200.0     15.99       147.64   
421066    20000.0      20000.0          20000.0      6.99       617.46   
421067    15000.0      15000.0          15000.0     11.99       498.15   
421068    13000.0      13000.0          13000.0      8.19       408.52   
421069    26950.0      26950.0          26950.0      8.19       846.88   
421070    15000.0      15000.0          15000.0     11.44       329.44   
421071     6000.0       6000.0           6000.0     17.86       216.50   
421072    18700.0      18700.0          18700.0     17.86       473.44   
421073    23000.0      23000.0          23000.0     15.59       554.32   
421074    25000.0      25000.0          25000.0     15.99       607.82   
421075    25000.0      25000.0          25000.0     15.59       602.52   
421076    26500.0      26500.0          26500.0      8.67       545.87   
421077     8000.0       8000.0           8000.0      8.19       251.40   
421078    21000.0      21000.0          21000.0      6.03       639.15   
421079    12000.0      12000.0          12000.0      9.49       384.34   
421080    10775.0      10775.0          10775.0      6.03       327.95   
421081     7000.0       7000.0           7000.0     14.31       240.30   
421082     6225.0       6225.0           6225.0     16.49       220.37   
421083    13150.0      13150.0          13150.0     14.99       312.77   
421084    10000.0      10000.0          10000.0      6.49       306.45   
421085     4000.0       4000.0           4000.0      8.67       126.59   
421086    10850.0      10850.0          10850.0     19.24       399.04   
421087     7

In [9]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.98993209, 0.98926723, 0.99050152, 0.99168824, 0.99183053,
       0.99128432, 0.99104683, 0.9930892 , 0.99301779, 0.99259031])

In [11]:
from sklearn.decomposition import PCA
pca = PCA(n_components=50)
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents)
principalDf.head()

0             1             2             3             4   \
0  263636.671395 -10322.591249  51285.884562  28905.804763  17712.380870   
1  330915.170803 -16161.293833 -15974.124774 -22332.295185  36524.432733   
2   55892.560597 -36618.413858  48309.198473 -24038.123409  -3414.921770   
3   -7629.499733 -39068.147755  49308.584863 -28607.543598   4512.080931   
4  148488.673572 -29968.962705  29192.600840  89613.658027  37403.238644   

             5             6             7             8             9   \
0 -34738.204272 -16550.666965  -5066.540421   1809.417565  17422.153455   
1  21241.397545  10950.395440  -9836.413271   -259.886666   2301.382269   
2 -17079.484709  13630.897056   3145.479804  10534.313334   2381.427102   
3  22900.119892   2474.680826   1588.568642   3171.273143  -1617.221663   
4 -18038.901538 -27512.189695  11818.524320   -686.597370   8784.428730   

     ...           40        41        42        43        44        45  \
0    ...    -0.802498  0.303507 -0.197150  0.835258  0.214588 -0.561621   
1    ...     1.105552  0.852156 -0.504651 -1.042981  0.853926 -0.876549   
2    ...     0.543002  0.083707 -0.379072 -0.112015  0.155520  0.588009   
3    ...    -0.607802  0.078878 -0.513208 -0.175742  0.780716  0.495183   
4    ...     1.052870  2.686966 -0.810383 -1.131298  0.734263  0.437996   

         46        47        48        49  
0 -0.545836  0.672934  0.365100 -0.302020  
1 -0.051822 -0.341286 -0.266565  0.569474  
2 -0.432801 -0.593755 -0.350122 -0.230913  
3 -0.790462  0.560436 -0.458171 -0.869700  
4 -0.582674  0.610101 -0.462856 -1.047329  

[5 rows x 50 columns]

In [12]:
X_pca_10 = principalDf
Y = y2015['loan_status']
X_pca_10 = pd.get_dummies(X_pca_10)
X_pca_10 = X_pca_10.dropna(axis=1)

cross_val_score(rfc, X_pca_10, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.98648905, 0.98786627, 0.98736702, 0.98715239, 0.98760331,
       0.98833951, 0.98800703, 0.98850575, 0.98750802, 0.98586933])

In [13]:
pca = PCA(n_components=40)
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents)

In [14]:
X_pca_40 = principalDf
Y = y2015['loan_status']
X_pca_40 = pd.get_dummies(X_pca_40)
X_pca_40 = X_pca_40.dropna(axis=1)

cross_val_score(rfc, X_pca_40, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.98736762, 0.98817495, 0.98736702, 0.9875561 , 0.98755581,
       0.98829201, 0.98784079, 0.98869573, 0.98760301, 0.98596433])

In [15]:
pca = PCA(n_components=30)
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents)

In [16]:
X_pca_30 = principalDf
Y = y2015['loan_status']
X_pca_30 = pd.get_dummies(X_pca_30)
X_pca_30 = X_pca_30.dropna(axis=1)

cross_val_score(rfc, X_pca_30, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.98691646, 0.98876858, 0.98836436, 0.98731863, 0.98810202,
       0.98814952, 0.98798328, 0.98855324, 0.98795925, 0.98648681])

In [17]:
pca = PCA(n_components=20)
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents)

In [18]:
X_pca_20 = principalDf
Y = y2015['loan_status']
X_pca_20 = pd.get_dummies(X_pca_20)
X_pca_20 = X_pca_20.dropna(axis=1)

cross_val_score(rfc, X_pca_20, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.98644156, 0.98838866, 0.98767572, 0.98798357, 0.98793578,
       0.98824451, 0.98824451, 0.98855324, 0.98748427, 0.98703303])

In [19]:
pca = PCA(n_components=10)
principalComponents = pca.fit_transform(X)
principalDf = pd.DataFrame(data = principalComponents)

In [20]:
X_pca_10 = principalDf
Y = y2015['loan_status']
X_pca_10 = pd.get_dummies(X_pca_10)
X_pca_10 = X_pca_10.dropna(axis=1)

cross_val_score(rfc, X_pca_10, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.72353612, 0.72341739, 0.72613507, 0.72670925, 0.72575283,
       0.72513537, 0.72660777, 0.72050442, 0.721139  , 0.72251645])

## The average accuracy dropped from ~99% to 72% somewhere between 20 to 30 components in PCA.